rskj-core/src/main/java/co/rsk/core/bc/ParallelizeTransactionHandler.java

```diff
@@ -0,0 +1,249 @@
+        long txGasLimit = GasCost.toGas(tx.getGasLimit());
+        return this.sequentialBucket.hasGasAvailable(txGasLimit);
+    }
+
+    public Optional<Short> addTransaction(Transaction tx, Set<ByteArrayWrapper> newReadKeys, Set<ByteArrayWrapper> newWrittenKeys, long gasUsedByTx) {
+        Set<Short> bucketCandidates = getBucketCandidates(tx, newReadKeys, newWrittenKeys);
+        long txGasLimit = GasCost.toGas(tx.getGasLimit());
+
+        if (bucketCandidates.size() > 1) {
+            return addTransactionInSequentialBucketAndGetId(tx, newReadKeys, newWrittenKeys, gasUsedByTx, txGasLimit);
+        }
+
+        if (bucketCandidates.size() == 1) {
+            Short bucketIdNumber = bucketCandidates.iterator().next();
+            TransactionBucket transactionBucket = this.parallelBuckets.get(bucketIdNumber);
+
+            if (!transactionBucket.hasGasAvailable(txGasLimit)) {
+                return addTransactionInSequentialBucketAndGetId(tx, newReadKeys, newWrittenKeys, gasUsedByTx, txGasLimit);
+            }
+
+            addTransactionInternal(tx, transactionBucket, gasUsedByTx, bucketIdNumber, newReadKeys, newWrittenKeys);
+            return Optional.of(bucketIdNumber);
+        }
+
+        Optional<Short> bucketId = getAvailableBucketWithLessUsedGas(txGasLimit);
+
+        if (!bucketId.isPresent()) {
+            return addTransactionInSequentialBucketAndGetId(tx, newReadKeys, newWrittenKeys, gasUsedByTx, txGasLimit);
+        }
+
+        Short bucketIdNumber = bucketId.get();
+        addTransactionInternal(tx, this.parallelBuckets.get(bucketIdNumber), gasUsedByTx, bucketIdNumber, newReadKeys, newWrittenKeys);
+        return bucketId;
+    }
+
+    public Optional<Short> addRemascTransaction(Transaction tx, Set<ByteArrayWrapper> newReadKeys, Set<ByteArrayWrapper> newWrittenKeys, long gasUsedByTx) {
+        return addTransactionInSequentialBucketAndGetId(tx, newReadKeys, newWrittenKeys, gasUsedByTx, GasCost.toGas(tx.getGasLimit()));
+    }
+
+    public long getGasUsedIn(Short bucketId) {
+        //TODO(Juli): If the bucketId doesnt exist, should it throw an error?
+        if (bucketId == this.sequentialBucket.getId()) {
+            return this.sequentialBucket.getGasUsed();
+        }
+
+        return this.parallelBuckets.get(bucketId).getGasUsed();
+    }
+
+    public List<Transaction> getTransactionsInOrder() {
+        List<Transaction> txs = new ArrayList<>();
+        for (TransactionBucket bucket: this.parallelBuckets) {
+            txs.addAll(bucket.getTransactions());
+        }
+
+        txs.addAll(this.sequentialBucket.getTransactions());
+        return txs;
+    }
+
+    public short[] getTransactionsPerBucketInOrder() {
+        List<Short> bucketSizes = new ArrayList<>();
+        short bucketEdges = 0;
+
+        for (TransactionBucket bucket: this.parallelBuckets) {
+            if (bucket.getTransactions().isEmpty()) {
+                break;
```

Comment:
shouldn't this be a `continue` instead of a `break`?

Reply:
Indeed, `break` was because the selection of the next available bucket was given the next available instead of the one with less gas.